### ISSUE TITLE
serialize fnIndexes in ArcadiaBehaviour

### ISSUE THREAD

**Issue by selfsame** [arcadia-unity/Arcadia#226]
slightly tricky:
* during serialization, non var fns are discarded
* before serializing the `fnIndexes` map, these discarded fns should be `dissoc`'d and the remaining entries reindexed

**Comment by nasser** [comment]
I wonder if a better internal representation for hooks might be a hashmap. We can regenerate the `IFn` array we need for fast runtime dispatch from the hashmap's `vals` every time a function is added or removed. Then, all we'd have to do is serialize the hashmap as an edn string after `dissoc`ing the non-var keys? The `qualifiedVarNames` variable strikes me as redundant.

**Comment by selfsame** [comment]
you'd still need to serialize the keys to strings no?  I'm also a bit worried about how 884a806d0bb312be42979a0d85a5d47bb28bbf15 works without a dedicated `fns` array, since `(hook+ o :update #'two)` would overwrite `(hook+ o :update #'one)`

**Comment by timsgardner** [comment]
agreed.

While we're at it, this is another not-at-all-threadsafe thing that uses persistent data structures anyway. 
All of this might be simplified and made threadsafe by representing it as a record class in an Atom, like so:
```clojure
(defrecord HookState [key-fns fn-arr])

(defn build-hook-state [key-fns]
  (StateState. key-fns (into-array (vals key-fns))))

(defn add-fn [state-component key f]
  (swap! (get-state-atom state-component)
    (fn [^HookState hs]
      (build-hook-state (assoc (.key-fns hs) key f)))))

(defn remove-fn [state-component key]
  (swap! (get-staet-atom state-component key f)
    (fn [^HookState hs]
      (build-hook-state (dissoc (.key-fns hs) key)))))
```

**Comment by timsgardner** [comment]
proposal:
https://github.com/arcadia-unity/Arcadia/commit/23b7fb6d03b6628971ab6665666e49f594e42b05

**Comment by timsgardner** [comment]
closed as of 5d1ba134e1ac384569348593f79d3a1223826361
